hypertrace-trace-enricher/trace-reader/src/main/java/org/hypertrace/trace/reader/attributes/DefaultValueResolver.java

```diff
@@ -72,108 +73,106 @@ private Single<LiteralValue> resolveDefinition(
-        .filter(literalValue -> !literalValue.getValueCase().equals(ValueCase.VALUE_NOT_SET))
-        .onErrorComplete();
-  }
-
-  private Single<LiteralValue> resolveValue(
+  private Optional<LiteralValue> resolveValue(
       ValueSource contextValueSource,
       String attributeScope,
       AttributeType attributeType,
       AttributeKind attributeKind,
       String path) {
-    Single<ValueSource> matchingValueSource =
-        Maybe.fromOptional(contextValueSource.sourceForScope(attributeScope))
-            .switchIfEmpty(
-                this.buildAndLogErrorLazily(
-                    "No value source available supporting scope %s", attributeScope));
+    Optional<ValueSource> matchingValueSource = contextValueSource.sourceForScope(attributeScope);
+    if (matchingValueSource.isEmpty()) {
+      if (LOGGING_LIMITER.tryAcquire()) {
+        log.error("No value source available supporting scope %s", attributeScope);
+      }
+      return Optional.empty();
+    }
     switch (attributeType) {
       case ATTRIBUTE:
-        return matchingValueSource
-            .mapOptional(valueSource -> valueSource.getAttribute(path, attributeKind))
-            .defaultIfEmpty(LiteralValue.getDefaultInstance());
+        return matchingValueSource.flatMap(
+            valueSource -> valueSource.getAttribute(path, attributeKind));
       case METRIC:
-        return matchingValueSource
-            .mapOptional(valueSource -> valueSource.getMetric(path, attributeKind))
-            .defaultIfEmpty(LiteralValue.getDefaultInstance());
+        return matchingValueSource.flatMap(
+            valueSource -> valueSource.getMetric(path, attributeKind));
       case UNRECOGNIZED:
       case TYPE_UNDEFINED:
       default:
-        return this.buildAndLogErrorLazily("Unrecognized projection type");
+        return Optional.empty();
     }
   }
 
-  private Single<LiteralValue> resolveProjection(ValueSource valueSource, Projection projection) {
+  private Optional<LiteralValue> resolveProjection(ValueSource valueSource, Projection projection) {
     switch (projection.getValueCase()) {
       case ATTRIBUTE_ID:
-        return valueSource
-            .executionContext()
-            .wrapSingle(() -> this.attributeClient.get(projection.getAttributeId()))
+        return this.attributeClient
+            .getById(valueSource.requestContext(), projection.getAttributeId())
             .flatMap(attributeMetadata -> this.resolve(valueSource, attributeMetadata));
       case LITERAL:
-        return Single.just(projection.getLiteral());
+        return Optional.of(projection.getLiteral());
       case EXPRESSION:
         return this.resolveExpression(valueSource, projection.getExpression());
       case VALUE_NOT_SET:
       default:
-        return this.buildAndLogErrorLazily("Unrecognized projection type");
+        if (LOGGING_LIMITER.tryAcquire()) {
+          log.error("Unrecognized projection type");
```

Comment:
pre-existing, but would be useful to actually log the projection we can't recognize